### PR TITLE
[REEF-803] Add a ConfigurationProvider for HadoopFileSystem

### PR DESCRIPTION
**Pull request by jwang98052**
Added HadoopFileSystemConfigurationProvider
Moved DriverConfigurationProviders  to Common as it is shared by multiple providers

JIRA: [REEF-803](https://issues.apache.org/jira/browse/REEF-803)

This closes #

### CODE REVIEW

**Comment by markusweimer** [comment]
I'll do a pass.

File: lang/cs/Org.Apache.REEF.Common/Io/TcpPortConfigurationProvider.cs [hunk 1]
```diff
@@ -17,12 +17,10 @@
  * under the License.
  */
 
-using System;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/TcpPortConfigurationProvider.cs]: The changes in this file, while net positive, seem to be orthogonal to the goals of the PR.

File: lang/cs/Org.Apache.REEF.Common/Parameters/DriverConfigurationProviders.cs [hunk 2]
```diff
@@ -21,10 +21,10 @@
 using Org.Apache.REEF.Tang.Annotations;
 using Org.Apache.REEF.Tang.Interface;
 
-namespace Org.Apache.REEF.Client.API.Parameters
+namespace Org.Apache.REEF.Common.Parameters
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Parameters/DriverConfigurationProviders.cs]: The namespace should still reflect the fact that this is a parameter on the client side. Maybe `Common.Client.Parameters`?

File: lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs [hunk 3]
```diff
@@ -72,9 +73,11 @@ internal static IConfiguration GetYarnIMRUConfiguration()
             IConfiguration imruClientConfig =
                 REEFIMRUClientConfiguration<TMapInput, TMapOutput, TResult>.ConfigurationModule.Build();
 
+            var fileSystemConfig = HadoopFileSystemConfiguration.ConfigurationModule.Build();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs]: This assumes that IMRU on REEF will always use this particular file system. While this is true in the here and now, it doesn't seem like a defensible position long term. Instead, we should requite the user to supply this configuration. Maybe {{IMRUJobDefinition}} needs a new field for {{DriverConfiguration}}?

File: lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HadoopFileSystemConfigurationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+using Org.Apache.REEF.Tang.Util;
+
+namespace Org.Apache.REEF.IO.FileSystem.Hadoop
+{
+    /// <summary>
+    /// This provider provides configuration for HardoopFileSystem
+    /// The client that is going to use HadoopFileSystem in its driver and evaluators should set 
+    /// configuration data through HadoopFileSystemConfiguration module in he client's configuration
+    /// </summary>
+    public class HadoopFileSystemConfigurationProvider : IConfigurationProvider
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HadoopFileSystemConfigurationProvider.cs]: `internal sealed`

**Comment by markusweimer** [comment]
@jwang98052 I have done a pass.

File: lang/cs/Org.Apache.REEF.Common/Io/TcpPortConfigurationProvider.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/TcpPortConfigurationProvider.cs]: put back.

File: lang/cs/Org.Apache.REEF.Common/Parameters/DriverConfigurationProviders.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Parameters/DriverConfigurationProviders.cs]: Done

File: lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs]: The code is put in the case of run on Yarn. And it is just an example. Clients can follow it to provide their own configurations.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HadoopFileSystemConfigurationProvider.cs (hunk 4 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HadoopFileSystemConfigurationProvider.cs]: done

**Comment by jwang98052** [comment]
review comments have been addressed and checked in.

File: lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs (hunk 3 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU.Examples/OnREEFIMRURunTimeConfiguration.cs]: Ah, OK.

**Comment by markusweimer** [comment]
This looks good. Will test and merge.
